nomos-services/data-availability/sampling/src/backend/kzgrs.rs

```diff
@@ -1,2 +1,397 @@
+    fn prune_by_time(&mut self) {
+        self.pending_sampling_blobs.retain(|_blob_id, context| {
+            context.started.elapsed() < self.settings.blobs_validity_duration
+        });
+    }
+}
+
+// TODO there is no logging at all here. Should we not do some logging?
+// Or can we assume that it is done on an upeer level by the clients?
+#[async_trait::async_trait]
+impl<R: Rng + Sync + Send> DaSamplingServiceBackend<R> for KzgrsDaSampler<R> {
+    type Settings = KzgrsDaSamplerSettings;
+    type BlobId = BlobId;
+    type Blob = DaBlob;
+
+    fn new(settings: Self::Settings, rng: R) -> Self {
+        let bt: BTreeSet<BlobId> = BTreeSet::new();
+        Self {
+            settings,
+            validated_blobs: bt,
+            pending_sampling_blobs: HashMap::new(),
+            rng,
+        }
+    }
+
+    async fn next_prune_interval(&self) -> Interval {
+        time::interval(self.settings.old_blobs_check_interval)
+    }
+
+    async fn get_validated_blobs(&self) -> BTreeSet<Self::BlobId> {
+        self.validated_blobs.clone()
+    }
+
+    async fn mark_in_block(&mut self, blobs_ids: &[Self::BlobId]) {
+        for id in blobs_ids {
+            self.pending_sampling_blobs.remove(id);
+            self.validated_blobs.remove(id);
+        }
+    }
+
+    async fn handle_sampling_success(&mut self, blob_id: Self::BlobId, blob: Self::Blob) {
+        if let Some(ctx) = self.pending_sampling_blobs.get_mut(&blob_id) {
+            ctx.subnets.insert(blob.column_idx as SubnetworkId);
+
+            println!("{}", self.settings.num_samples);
+            println!("{}", ctx.subnets.len());
+            match ctx.subnets.len() {
+                // sampling of this blob_id terminated successfully
+                len if len == self.settings.num_samples as usize => {
+                    self.validated_blobs.insert(blob_id);
+                    // cleanup from pending samplings
+                    self.pending_sampling_blobs.remove(&blob_id);
+                }
+                len if len > self.settings.num_samples as usize => {
+                    unreachable!("{}", "more subnets than expected after sampling success!");
+                }
+                // do nothing if smaller
+                _ => {}
+            }
+        } else {
+            unreachable!("{}", NON_TRIGGERED_ERR);
+        }
+    }
+
+    async fn handle_sampling_error(&mut self, blob_id: Self::BlobId) {
+        // If it fails a single time we consider it failed.
+        // We may want to abstract the sampling policies somewhere else at some point if we
+        // need to get fancier than this
+        self.pending_sampling_blobs.remove(&blob_id);
+        self.validated_blobs.remove(&blob_id);
+    }
+
+    // TODO: This also would be an error which should never happen, but what if a client starts
+    // init_sampling of a blob which is already pending? Or worse, which already is validated?
+    // Should we not therefore return an error here?
```

Comment:
Imo this object should be **idempotent**. Why?, Because we cannot ensure that a single service will try to start sampling or check how things are going (even if for now it is the case). So, maybe here we could add a check and just start sampling if it wasn't pending already. wdyt?

Reply:
Currently, if already sampling, this would result in the network adapter resending SampleRequest messages. Out of band you suggested an `enum` return signaling the state, so that's what I did here.